tidy3d/components/eme/simulation.py

```diff
@@ -0,0 +1,335 @@
+from ..structure import Structure
+from ..medium import Medium
+from ..viz import plot_params_structure
+from ..geometry.base import Box
+from ...exceptions import ValidationError
+from ...log import log
+from ...constants import C_0
+
+from .monitor import EMEMonitorType
+from .grid import EMEGridSpecType, EMEGrid
+
+
+# dummy run_time for converting to FDTD simulation
+# should be fairly small, or else we run into issues constructing tmesh
+RUN_TIME = 1e-30
+
+
+class EMESimulation(AbstractSimulation):
+    """EME simulation.
+    An EME simulation always computes the full scattering matrix of the structure.
+    Additional data can be recorded by adding 'monitors' to the simulation.
+
+    Example
+    -------
+    >>> from tidy3d import Box, Medium, Structure, C_0, inf
+    >>> from tidy3d import ModeSpec, EMEUniformGrid, GridSpec
+    >>> from tidy3d import EMEFieldMonitor
+    >>> lambda0 = 1
+    >>> freq0 = C_0 / lambda0
+    >>> sim_size = 3*lambda0, 3*lambda0, 3*lambda0
+    >>> waveguide_size = (lambda0/2, lambda0, inf)
+    >>> waveguide = Structure(
+    ...     geometry=Box(center=(0,0,0), size=waveguide_size),
+    ...     medium=Medium(permittivity=2)
+    ... )
+    >>> eme_grid_spec = EMEUniformGrid(num_cells=5, mode_spec=ModeSpec(num_modes=10))
+    >>> grid_spec = GridSpec(wavelength=lambda0)
+    >>> field_monitor = EMEFieldMonitor(
+    ...     size=(0, sim_size[1], sim_size[2]),
+    ...     name="field_monitor"
+    ... )
+    >>> sim = EMESimulation(
+    ...     size=sim_size,
+    ...     monitors=[field_monitor],
+    ...     structures=[waveguide],
+    ...     axis=2,
+    ...     freqs=[freq0],
+    ...     eme_grid_spec=eme_grid_spec,
+    ...     grid_spec=grid_spec
+    ... )
+    """
+
+    freqs: List[pd.PositiveFloat] = pd.Field(
+        ..., title="Frequencies", description="Frequencies for the EME simulation.", min_items=1
+    )
+
+    axis: Axis = pd.Field(
+        ...,
+        title="Propagation Axis",
+        description="Propagation axis (0, 1, or 2) for the EME simulation.",
+    )
+
+    eme_grid_spec: EMEGridSpecType = pd.Field(
+        ...,
+        title="EME Grid Specification",
+        description="Specification for the EME propagation grid. "
+        "The simulation is divided into cells in the propagation direction; "
+        "this parameter specifies the layout of those cells. "
+        "Mode solving is performed in each cell, and then propagation between cells "
+        "is performed to determine the complete solution. "
+        "This is distinct from 'grid_spec', which defines the grid in the two "
+        "tangential directions, as well as the grid used for field monitors.",
+    )
+
+    monitors: Tuple[EMEMonitorType, ...] = pd.Field(
+        (),
+        title="Monitors",
+        description="Tuple of monitors in the simulation. "
+        "Note: monitor names are used to access data after simulation is run.",
+    )
+
+    boundary_spec: BoundarySpec = pd.Field(
```

Comment:
I will probably disallow boundary_spec and sources. Boundary conditions are incorporated instead in ModeSpec, and sources that aren't in the EME modes will be handled in the smatrix plugin.